service/worker/failoverManager/workflow.go

```diff
@@ -0,0 +1,376 @@
+
+	// FailoverResult is workflow result
+	FailoverResult struct {
+		SuccessDomains []string
+		FailedDomains  []string
+	}
+
+	// GetDomainsActivityParams params for activity
+	GetDomainsActivityParams struct {
+		TargetCluster string
+		SourceCluster string
+		Domains       []string
+	}
+
+	// FailoverActivityParams params for activity
+	FailoverActivityParams struct {
+		Domains       []string
+		TargetCluster string
+	}
+
+	// FailoverActivityResult result for failover activity
+	FailoverActivityResult struct {
+		SuccessDomains []string
+		FailedDomains  []string
+	}
+
+	// QueryResult ..
+	QueryResult struct {
+		TotalDomains   int
+		Success        int
+		Failed         int
+		State          string
+		TargetCluster  string
+		SourceCluster  string
+		SuccessDomains []string
+		FailedDomains  []string
+	}
+)
+
+func init() {
+	workflow.RegisterWithOptions(FailoverWorkflow, workflow.RegisterOptions{Name: WorkflowTypeName})
+	activity.RegisterWithOptions(FailoverActivity, activity.RegisterOptions{Name: failoverActivityName})
+	activity.RegisterWithOptions(GetDomainsActivity, activity.RegisterOptions{Name: getDomainsActivityName})
+}
+
+// FailoverWorkflow is the workflow that managed failover all domains with IsManagedByCadence=true
+func FailoverWorkflow(ctx workflow.Context, params *FailoverParams) (*FailoverResult, error) {
+	err := validateParams(params)
+	if err != nil {
+		return nil, err
+	}
+
+	// define query properties
+	var failedDomains []string
+	var successDomains []string
+	var totalNumOfDomains int
+	wfState := wfRunning
+	err = workflow.SetQueryHandler(ctx, QueryType, func(input []byte) (*QueryResult, error) {
+		return &QueryResult{
+			TotalDomains:   totalNumOfDomains,
+			Success:        len(successDomains),
+			Failed:         len(failedDomains),
+			State:          wfState,
+			TargetCluster:  params.TargetCluster,
+			SourceCluster:  params.SourceCluster,
+			SuccessDomains: successDomains,
+			FailedDomains:  failedDomains,
+		}, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// get target domains
+	ao := workflow.WithActivityOptions(ctx, getGetDomainsActivityOptions())
+	getDomainsParams := &GetDomainsActivityParams{
+		TargetCluster: params.TargetCluster,
+		SourceCluster: params.SourceCluster,
+		Domains:       params.Domains,
+	}
+	var domains []string
+	err = workflow.ExecuteActivity(ao, GetDomainsActivity, getDomainsParams).Get(ctx, &domains)
+	if err != nil {
+		return nil, err
+	}
+	totalNumOfDomains = len(domains)
+
+	// failover in batch
+	ao = workflow.WithActivityOptions(ctx, getFailoverActivityOptions())
+	batchSize := params.BatchFailoverSize
+	times := len(domains) / batchSize
+
+	pauseCh := workflow.GetSignalChannel(ctx, PauseSignal)
+	resumeCh := workflow.GetSignalChannel(ctx, ResumeSignal)
+	var shouldPause bool
+
+	for i := 0; i < times; i++ {
+		// check if need to pause
+		shouldPause = pauseCh.ReceiveAsync(nil)
+		if shouldPause {
+			wfState = wfPaused
+			resumeCh.Receive(ctx, nil)
+		}
```

Comment:
so we don't plan to handle the case where resume signal is sent before pause or multiple pause signal comes in before a resume signal?

Reply:
My current thinking with this simple pause&resume implementation is operator can always find a way out if they mistakenly pause or resume. 
Another benefit arguable is you can achieve pause twice then resume once to achieve some rate control

I am open to other suggestions.

Reply:
From the CLI, it can check the workflow state before sending a resume signal